### PR TITLE
Untangle UserDefinedFunctionService from NodeContext/Schemas

### PR DESCRIPTION
**Pull request by mfussenegger**
Instead of having each `SchemaInfo` instance listen to cluster state
events and filter the UDF metadata by schemaName, to then trigger an
update on the `Functions`, this promotes the
`UserDefinedFunctionService` to a lifecycle
component/ClusterStateListener that always updates UDFs for _all_
schemas.

This avoids the need for the filtering, and lets us replace the
`ConcurrentHashMap` with a regular map that's volatile and gets replaced
on a update.

It also makes the `NodeContext` bootstrapping a bit easier and lets us
share more code between tests and production.

### CODE REVIEW

File: server/src/main/java/io/crate/metadata/NodeContext.java [hunk 1]
```diff
@@ -21,16 +21,61 @@
...
+                                 TableStats tableStats) {
+        return new NodeContext(functions, roles, nodeCtx -> {
+            var tableInfoFactory = new DocTableInfoFactory(nodeCtx);
+            BlobSchemaInfo blobSchemaInfo = new BlobSchemaInfo(
+                clusterService,
+                new BlobTableInfoFactory(clusterService.getSettings(), environment));
+            Map<String, SchemaInfo> systemSchemas = Map.of(
+                SysSchemaInfo.NAME, new SysSchemaInfo(clusterService, nodeCtx.roles()),
+                InformationSchemaInfo.NAME, new InformationSchemaInfo(),
+                PgCatalogSchemaInfo.NAME, new PgCatalogSchemaInfo(tableStats, nodeCtx.roles()),
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/NodeContext.java:63]: In a followup it should be possible to remove the `tableStats` parameter too, and then the systemSchemas definition could also move into the `Schemas` constructor.

File: server/src/test/java/io/crate/metadata/doc/DocSchemaInfoTest.java [hunk 2]
```diff
@@ -107,34 +103,11 @@ public String name() {
...
             clusterService,
-            nodeCtx,
-            udfService,
             new ViewInfoFactory(() -> null),
             new DocTableInfoFactory(nodeCtx)
         );
     }
 
-    @Test
-    public void testInvalidFunction() throws Exception {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/metadata/doc/DocSchemaInfoTest.java:118]: Moved this as ` test_cannot_create_invalid_function` to `UserDefinedFunctionsTest`

Comment (seut) [reply server/src/test/java/io/crate/metadata/doc/DocSchemaInfoTest.java:118]: Looks like the `udfService` is then no longer needed in this test suite?

**Review by seut — APPROVED** [review]
Nice! Just a minor suggestion.

**Pull request merged**